### ISSUE TITLE
Random NaN loss when using float16 dtype and batch size of 1

### ISSUE THREAD

**Issue by PaulPauls** [tensorflow/tensorflow#43172]
**System information**
- OS Platform and Distribution: Google Colab
- TensorFlow version (use command below): tf-nightly (2.4.0-dev20200911)

**Reproducing code**
see [https://colab.research.google.com/drive/1miE61h7rlSJL0xIG4yNdVx05lVxT-oKb?usp=sharing](https://colab.research.google.com/drive/1miE61h7rlSJL0xIG4yNdVx05lVxT-oKb?usp=sharing)

````
import math
import numpy as np
import tensorflow as tf

print(tf.__version__)

x = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
y = np.array([[0], [1], [1], [0]])

loss_function = tf.keras.losses.BinaryCrossentropy()

for i in range(2000):
    if i % 100 == 0:
        print("Iteration {}".format(i))

    model = tf.keras.Sequential([
        tf.keras.layers.Dense(units=1, dtype=tf.float16)])

    model.compile(optimizer='sgd', loss=loss_function)
    model.fit(x=x, y=y, epochs=1, batch_size=1, verbose=0)

    loss_result = loss_function(y, model(x))

    if math.isnan(loss_result):
        raise RuntimeError('NAN Error in iteration {}'.format(i))
````

**Behavior description**

Seemingly non-deterministic occurence of a NaN result when calculating loss of a very simple Dense Model. The NaN loss seems to happen randomly and can occur on the 60th or 600th iteration. In the supplied Google colab code it happened in the 343rd iteration. The bug only seems to occur using a dtype of float16 and batch_size of 1. Debugging the error lead me to see that the models producing a NaN loss seem to have been initialized with a NaN bias and kernel, though I couldn't get to the bottom of why.

**Background**

I already opened this issue before in the following ticket: [https://github.com/tensorflow/tensorflow/issues/38457](https://github.com/tensorflow/tensorflow/issues/38457)

The assigned Tensorflower closed the issue after supplying a hotfix, with the note that I shall reopen the issue if I am not satisfied with the solution. As the supplied solution is only a hotfix, though does not solve the underlying issue even in the current tf nightly (see the Google colab above in which I tested the current nightly) am I not satisfied with the solution, particularly since the issue is very easily reproducible by setting the dtype to float16 and batch_size to 1. The ticket was closed with the request to reopen the issue, if not satisfied, which I understood as the request to create a new issue as I am obviously no TF contributor and can't reopen issues...

**Comment by bhack** [comment]
This is an `edge` overflow initialization extracted from your loop:

Try to run this multiple time this and e.g. change it with `tf.random.set_seed(1)` or removing the seed.
```
tf.random.set_seed(0)
x = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
y = np.array([[0], [1], [1], [0]])

loss_function = tf.keras.losses.BinaryCrossentropy()

l=[]
w=np.array([[-0.2373],[ 1.    ]], dtype=np.float16) #weights
b=np.array([0.], dtype=np.float16) #array of biases
l.append(w)
l.append(b)

model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(units=1, input_dim=2, dtype=tf.float16))

opt = tf.keras.optimizers.SGD(learning_rate=0.1)
model.layers[0].set_weights(l)
print("Init weights")
print(model.get_weights())
model.compile(optimizer=opt, loss=loss_function) 
model.fit(x=x, y=y, epochs=1, batch_size=1, verbose=0)
print("Fitted weights")
print(model.get_weights())

loss_result = loss_function(y, model(x, training=False))

if math.isnan(loss_result):
    print('NAN Error')
```

**Comment by bhack** [comment]
About the backward pass (in fit) see also https://www.tensorflow.org/guide/mixed_precision?hl=en#underflow_and_overflow

**Comment by PaulPauls** [comment]
Thank you very much @bhack for supplying the reasoning behind the bug!
Do I understand it correctly then that this bug is due to a mixed precision constraint by the hardware and fixing it won't be possible without slowing down the mixed precision processing significantly? Tensorflow therefore can't do anything about this issue!?

Is this a known constraint or how did you manage to debug the framework and find that out? I would love to know more about doing such complicated debugging to improve my skills with the in internals of Tensorflow. I would be thankful if you could point me to some good learning resources for such things if you know any.

Either way, thank you very much for the response! Also please excuse my late reply, as I was on a holiday last week.

**Comment by bhack** [comment]
If you want you can play a little bit with a custom loop:
https://www.tensorflow.org/guide/mixed_precision?hl=en#training_the_model_with_a_custom_training_loop

You can close the ticket, for other questions that are not Bugs or Feature Requests please use our channels:
https://stackoverflow.com/questions/tagged/tensorflow
https://www.tensorflow.org/community/forums
